components/core/src/BufferedFileReader.cpp

```diff
@@ -0,0 +1,385 @@
+void BufferedFileReader::open (const string& path) {
+    ErrorCode error_code = try_open(path);
+    if (ErrorCode_Success != error_code) {
+        if (ErrorCode_FileNotFound == error_code) {
+            SPDLOG_ERROR("File not found: {}", boost::filesystem::weakly_canonical(path).string());
+            throw OperationFailed(error_code, __FILENAME__, __LINE__);
+        } else {
+            throw OperationFailed(error_code, __FILENAME__, __LINE__);
+        }
+    }
+}
+
+void BufferedFileReader::close () {
+    if (-1 != m_fd) {
+        // NOTE: We don't check errors for fclose since it seems
+        // the only reason it could fail is if it was interrupted by a signal
+        ::close(m_fd);
+        m_fd = -1;
+
+        if (m_checkpoint_pos.has_value()) {
+            m_buffer_size = m_base_buffer_size;
+            m_buffer = make_unique<char[]>(m_buffer_size);
+            m_checkpoint_pos.reset();
+        }
+    }
+}
+
+ErrorCode BufferedFileReader::try_fstat (struct stat& stat_buffer) const {
+    if (-1 == m_fd) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    auto return_value = fstat(m_fd, &stat_buffer);
+    if (0 != return_value) {
+        return ErrorCode_errno;
+    }
+    return ErrorCode_Success;
+}
+
+size_t BufferedFileReader::set_checkpoint() {
+    if (m_checkpoint_pos.has_value() && m_checkpoint_pos < m_file_pos) {
+        if (m_buffer_reader->get_buffer_size() != m_base_buffer_size) {
+            // allocate new buffer for buffered data starting from pos
+            resize_buffer_from_pos(m_buffer_reader->get_pos());
+            m_buffer_reader->seek_from_begin(get_corresponding_offset(m_file_pos));
+        }
+    }
+    m_checkpoint_pos = m_file_pos;
+    return m_file_pos;
+}
+
+void BufferedFileReader::clear_checkpoint () {
+    if (false == m_checkpoint_pos.has_value()) {
+        return;
+    }
+
+    m_file_pos = highest_read_pos;
+    resize_buffer_from_pos(get_corresponding_offset(m_file_pos));
+    m_checkpoint_pos.reset();
+}
+
+ErrorCode BufferedFileReader::peek_buffered_data (size_t size_to_peek, const char*& data_ptr,
+                                                  size_t& peek_size) {
+    if (-1 == m_fd) {
+        return ErrorCode_NotInit;
+    }
+    // Refill the buffer if it is not loaded yet
+    if (false == m_buffer_reader.has_value()) {
+        auto error_code = refill_reader_buffer(m_base_buffer_size);
+        if (ErrorCode_Success != error_code) {
+            data_ptr = nullptr;
+            peek_size = 0;
+            return error_code;
+        }
+    }
+    m_buffer_reader->peek_buffer(size_to_peek, data_ptr, peek_size);
+    return ErrorCode_Success;
+}
+
+size_t BufferedFileReader::quantize_to_buffer_size (size_t size) const {
+    if (size == 0) {
+        return 0;
+    }
+    return (1 + ((size - 1) / m_base_buffer_size)) * m_base_buffer_size;
+}
+
+ErrorCode BufferedFileReader::refill_reader_buffer (size_t num_bytes_to_refill) {
```

Comment:
I feel like this method would be more clear with the following structure:
```
if checkpoint:
    grow the buffer if necessary
    calculate the number of bytes to read
    calculate the destination for the read data
        buf_begin_pos = m_buf_reader->get_pos()
else:
    if we haven't filled the current base-buffer-size:
        calculate the number of bytes to read
        calculate the destination for the read data
        buf_begin_pos = m_buf_reader->get_pos()
    else:
        advance the buffer's position
        set the number of bytes to read to base buffer size
        set the destination for the read data to 0
        buf_begin_pos = 0

read num_bytes_to_read from the file
if there's an error and it's not EOF:
    return the error

Initialize the buffer reader with the variables you calculated above

return the error_code
```

Reply:
refactored in a way similar to the structure you proposed but I am not sure if it is what you expect